samza-kafka/src/test/java/org/apache/samza/system/kafka/TestKafkaSystemConsumer.java

```diff
@@ -205,6 +215,156 @@ public void testFetchThresholdBytesDiabled() {
+
+    final StartpointUpcoming testStartpointSpecific = new StartpointUpcoming();
+
+    // Mock the consumer interactions.
+    Mockito.doNothing().when(consumer).seekToEnd(ImmutableList.of(testTopicPartition));
+
+    // Invoke the consumer with startpoint.
+    kafkaStartpointRegistrationHandler.visit(testSystemStreamPartition, testStartpointSpecific);
+
+    // Mock verifications.
+    Mockito.verify(consumer).seekToEnd(ImmutableList.of(testTopicPartition));
+  }
+
+  @Test
+  public void testStartInvocationAfterStartPointsRegistrationShouldInvokeTheStartPointApplyMethod() {
+    // Initialize the constants required for the test.
+    Consumer mockConsumer = Mockito.mock(Consumer.class);
+    String testSystemName = "testSystem";
+    String testStreamName = "testStream";
+    Config testConfig = new MapConfig();
+    String testClientId = "testClientId";
+    KafkaSystemConsumerMetrics kafkaSystemConsumerMetrics = new KafkaSystemConsumerMetrics(testSystemName, new NoOpMetricsRegistry());
+
+    // Test system stream partitions.
+    SystemStreamPartition testSystemStreamPartition1 = new SystemStreamPartition(testSystemName, testStreamName, new Partition(0));
+    SystemStreamPartition testSystemStreamPartition2 = new SystemStreamPartition(testSystemName, testStreamName, new Partition(1));
+    SystemStreamPartition testSystemStreamPartition3 = new SystemStreamPartition(testSystemName, testStreamName, new Partition(2));
+    SystemStreamPartition testSystemStreamPartition4 = new SystemStreamPartition(testSystemName, testStreamName, new Partition(3));
+
+    // Different kinds of {@code Startpoint}.
+    StartpointSpecific startPointSpecific = new StartpointSpecific("100");
+    StartpointTimestamp startpointTimestamp = new StartpointTimestamp(100L);
+    StartpointOldest startpointOldest = new StartpointOldest();
+    StartpointUpcoming startpointUpcoming = new StartpointUpcoming();
+
+    // Mock the visit methods of KafkaStartpointRegistrationHandler.
+    KafkaSystemConsumer.KafkaStartpointRegistrationHandler
+        mockStartPointVisitor = Mockito.mock(KafkaStartpointRegistrationHandler.class);
+    Mockito.doNothing().when(mockStartPointVisitor).visit(testSystemStreamPartition1, startPointSpecific);
+    Mockito.doNothing().when(mockStartPointVisitor).visit(testSystemStreamPartition2, startpointTimestamp);
+    Mockito.doNothing().when(mockStartPointVisitor).visit(testSystemStreamPartition3, startpointOldest);
+    Mockito.doNothing().when(mockStartPointVisitor).visit(testSystemStreamPartition4, startpointUpcoming);
+
+    // Instantiate KafkaSystemConsumer for testing.
+    KafkaSystemConsumer kafkaSystemConsumer = new KafkaSystemConsumer(mockConsumer, testSystemName, testConfig,
+                                                                      testClientId, kafkaSystemConsumerMetrics, new TestClock(), mockStartPointVisitor);
+    kafkaSystemConsumer.proxy = Mockito.mock(KafkaConsumerProxy.class);
```

Comment:
Could you please try to avoid directly accessing and mutating fields? Maybe include the proxy in your VisibleForTesting constructor.